functions/index.js

```diff
@@ -255,39 +267,103 @@ function startTimer(time, roomID, prevState, newState) {
       .then(totalTime => {
           return console.log('Timer of ' + totalTime + ' has finished.');
       })
-      .then(() => new Promise(resolve => setTimeout(resolve, 1000)))
       .then(() => admin.database().ref(parentRoomID + roomID + "/state").set(newState))
       .catch(error => console.error(error));
 }
 
-function createNode(roomID, node) {
+function startTimerWithNodeCreation(time, roomID, prevState, newState) {
+  return functionTimer(time, prevState, roomID, elapsedTime => {
+          return admin.database().ref(parentRoomID + roomID + "/timer/observableTime").set(elapsedTime);
+      })
+      .then(totalTime => {
+          return console.log('Timer of ' + totalTime + ' has finished.');
+      })
+      .then(() => createNode(roomID, "onlineStatus", 0))
+      .then(() => admin.database().ref(parentRoomID + roomID + "/state").set(newState))
+      .catch(error => console.error(error));
+}
+
+function startTimerWithOnlineCheck(time, roomID, prevState, newState) {
+  return functionTimer(time, prevState, roomID, elapsedTime => {
+          return admin.database().ref(parentRoomID + roomID + "/timer/observableTime").set(elapsedTime);
+      })
+      .then(totalTime => {
+          return console.log('Timer of ' + totalTime + ' has finished.');
+      })
+      .then(() => {return checkOnlineNode(roomID)})
+      .then(() => admin.database().ref(parentRoomID + roomID + "/state").set(newState))
+      .catch(error => console.error(error));
+}
+
+function createNode(roomID, node, value) {
   updates = {};
-  admin.database().ref(parentRoomID + roomID).child("users").once('value', (snapshot) => {
+  return admin.database().ref(parentRoomID + roomID).child("users").once('value', (snapshot) => {
     snapshot.forEach((child) => {
-      updates[child.val()] = 0;
+      updates[child.val()] = value;
+    });
+  }).then(() => { 
+    return admin.database().ref(parentRoomID + roomID).child(node).once('value', function(snapshot) {
+      snapshot.ref.set(updates);
     });
   });
-
-  admin.database().ref(parentRoomID + roomID).child(node).once('value', function(snapshot) {
-    snapshot.ref.set(updates);
-  });
-
-  return;
 }
 
-function updateUser(userID, child, position) {
+function updateUser(userID) {
   return admin.database().ref("users/").once('value', (snapshot) => {
     if(snapshot.hasChild(userID)) {
-      admin.database().ref("users/" + userID + "/stars").transaction((currentValue) => {
-          return currentValue + child.val();
-      });
-      admin.database().ref("users/" + userID + "/trophies").transaction((currentValue) => {
-          return currentValue + getRankFromPosition(position);
+      let newTrophies;
+      let totalMatches;
+
+      return admin.database().ref("users/" + userID + "/trophies").transaction((currentValue) => {
+        newTrophies = currentValue + minRank;
+        return newTrophies;
+      })
+      .then(() => {
```

Comment:
Is it possible to do all the updates in parallel in this case? Just asking out of curiosity, you don’t need to actually do it haha

Reply:
Yes you can and that's what I did at the beginning but for some reason there was a weird bug with this. But for example the league update needs to wait for the trophies update so they need to be sequential as well as the change average which needs to wait for the increase match. So we could potentially split in two.